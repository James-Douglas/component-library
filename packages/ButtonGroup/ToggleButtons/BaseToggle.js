import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { useId } from '@comparethemarketau/manor-hooks';
import { StyledToggle, StyledToggleInput } from './BaseToggle.styles';

const BaseToggle = ({
  id: propsId,
  value,
  name,
  selectedValue,
  invalid,
  disabled,
  handleToggle,
  handleFocus,
  handleBlur,
  handleClick,
  checkedFontColor,
  children,
  className,
  button,
  inputType,
}) => {
  const id = useId(propsId);
  const [checked, setChecked] = useState(selectedValue || false);
  const wrapperElement = useRef(null);

  useEffect(() => {
    setChecked(selectedValue === value);
  }, [selectedValue, value]);

  const changeHandler = (e) => {
    const targetChecked = e.target.checked;
    setChecked(targetChecked);
    handleToggle(value);
  };

  return (
    <StyledToggle
      invalid={invalid}
      ref={wrapperElement}
      className={className}
      checked={checked}
      button={button}
    >
      <StyledToggleInput
        tabIndex={0}
        invalid={invalid}
        id={id}
        type={inputType}
        onChange={changeHandler}
        onClick={handleClick}
        checked={checked}
        disabled={disabled}
        name={name}
        value={value}
        onFocus={handleFocus}
        onBlur={handleBlur}
        checkedFontColor={checkedFontColor}
        button={button}
      />
      {children}
    </StyledToggle>
  );
};

BaseToggle.propTypes = {
  /**
   * Unique identifier for the BaseToggle
   */
  id: PropTypes.string,
  /**
   * Value applied to the input field
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  /**
   * The value of the currently selected (toggled on) toggle.
   * (`ToggleGroup` will override any value provided to this prop if in use)
   */
  selectedValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * The name to be applied to the input field.
   */
  name: PropTypes.string,
  /**
   * Applies invalid styling when true.
   */
  invalid: PropTypes.bool,
  /**
   * Disables the toggle when true.
   */
  disabled: PropTypes.bool,
  /**
   * Handler function called when a toggle is toggled on with the value of the toggle.
   */
  handleToggle: PropTypes.func.isRequired,
  /**
   * Handler function called on focus of the toggle
   */
  handleFocus: PropTypes.func,
  /**
   * Handler function called on blur of the toggle
   */
  handleBlur: PropTypes.func,
  /**
   * Handler function called on click of the toggle
   */
  handleClick: PropTypes.func,
  /**
   * Used by the `Color` toggle to override the checked label's font color for when displaying
   * light colors where white text is difficult to discern.
   */
  checkedFontColor: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  /**
   * Used by the `Button` toggle to determine if the toggle needs to be styled like a button.
   */
  button: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  /**
   * The type attribute for the underlying input element
   */
  inputType: PropTypes.oneOf(['radio', 'checkbox']),
};

BaseToggle.defaultProps = {
  id: null,
  name: '',
  selectedValue: null,
  invalid: false,
  disabled: false,
  checkedFontColor: null,
  children: [],
  className: '',
  handleFocus: null,
  handleBlur: null,
  handleClick: null,
  button: null,
  inputType: 'radio',
};

export default BaseToggle;
