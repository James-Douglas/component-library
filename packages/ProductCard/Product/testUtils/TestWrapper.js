import React from 'react';
import ProductProvider from '../../Context/ProductProvider';

const ProductTestWrapper = ({
  // eslint-disable-next-line react/prop-types
  children, product, onDetails = () => {}, onSelect = () => {},
}) => (
  <ProductProvider product={product} onDetails={onDetails} onSelect={onSelect}>
    { children }
  </ProductProvider>
);

export default ProductTestWrapper;
