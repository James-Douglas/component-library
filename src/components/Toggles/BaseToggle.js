import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider, css } from 'styled-components';
import getTheme from 'utils/getTheme';

const StyledToggle = styled.div`
  display: flex;
  position: relative;
  cursor: pointer;
  border: ${({ theme, invalid }) => (invalid ? theme.borders.invalid : theme.borders.component)};
  font-weight: ${({ theme }) => theme.fontWeight.bold};
  margin: ${({ theme }) => `0 ${theme.spacing['8']} ${theme.spacing['16']} ${theme.spacing['8']}`};
  background: ${({ theme }) => theme.toggle.base.background};
  box-shadow: ${({ theme }) => theme.boxShadow.sm};
  line-height: ${({ theme }) => theme.lineHeight.normal};
  min-width: 12.8rem;
`;

const StyledToggleInput = styled.input`
  opacity: 0;
  width:0;
  height: 0;
  position: absolute;
  &:checked + label svg {
    color: ${({ theme }) => theme.toggle.base.background}};
    fill: currentColor;
    box-shadow: none;
     transition : all 200ms ease-out;
  } 
  &:disabled + label:hover {
    cursor: not-allowed;
  }
  &:disabled + label {
    color: ${({ theme }) => theme.colors.disabled};
    border: ${({ theme }) => theme.borders.transparent};
    fill: currentColor;
  }
  &:checked + label {
    background: ${({ theme }) => theme.toggle.base.backgroundChecked};
    color: ${({ theme }) => theme.toggle.base.color};
    box-shadow: none;
  }
  &:focus + label {
    box-shadow: 0 0 2px 3px rgba(0, 123, 255, .3);
  }
  ${({ invalid, theme }) => invalid && css`
     & + label {
      color: ${theme.colors.invalid};
      fill: currentColor;
     }
     &:checked + label {
      color:  ${theme.colors.invalid};
      fill: currentColor;
    }
  `}
`;

const columns = {
  1: '100%', 2: '50%', 3: '33.33%', 4: '25%', 5: '20%',
};

export function getInlineStyles(type, rectOptions) {
  if (type !== 'rectangle') return {};

  const { col, height, align } = rectOptions;

  const inlineStyles = {
    height: `${height}rem`,
    alignItems: `${align}`,
  };

  if (col > 0) {
    inlineStyles.width = columns[col];
    inlineStyles.flexBasis = columns[col];
  }

  return inlineStyles;
}

const BaseToggle = ({
  id, type, value, name, selectedValue, invalid, disabled, handleToggle, rectOptions, children, className,
}) => {
  const wrapperElement = useRef(null);
  const toggleElement = useRef(null);

  const toggleHandler = () => {
    handleToggle(value);
  };

  return (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions
    <ThemeProvider theme={getTheme()}>
      <StyledToggle
        style={getInlineStyles(type, rectOptions)}
        onClick={() => {
          toggleElement.current.click();
          toggleElement.current.focus();
        }}
        invalid={invalid}
        ref={wrapperElement}
        className={className}
      >
        <StyledToggleInput
          tabIndex={0}
          ref={toggleElement}
          invalid={invalid}
          id={id}
          type="radio"
          onChange={toggleHandler}
          checked={selectedValue === value}
          disabled={disabled}
          name={name}
          value={value}
        />
        {children}
      </StyledToggle>
    </ThemeProvider>

  );
};

BaseToggle.propTypes = {
  id: PropTypes.string.isRequired,
  type: PropTypes.oneOf(['square', 'rectangle', 'custom']).isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  selectedValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  name: PropTypes.string,
  invalid: PropTypes.bool,
  disabled: PropTypes.bool,
  handleToggle: PropTypes.func.isRequired,
  rectOptions: PropTypes.shape({
    align: PropTypes.oneOf(['center', 'left', 'right']),
    col: PropTypes.oneOf([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
    height: PropTypes.number,
  }),
  children: PropTypes.node,
  className: PropTypes.string,
};

BaseToggle.defaultProps = {
  name: '',
  selectedValue: null,
  invalid: false,
  disabled: false,
  rectOptions: {
    align: 'center',
    col: 1,
    height: 8,
  },
  children: [],
  className: '',
};

export default BaseToggle;
