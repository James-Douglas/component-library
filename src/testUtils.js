// test-utils.js
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './providers/ManorProvider';

const customRender = (ui, options) => render(ui, { wrapper: ManorProvider, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
