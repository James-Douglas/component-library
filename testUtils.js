import React from 'react';
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './packages/Provider/ManorProvider';

const ManorTestWrapper = ({ children }) => (
  <ManorProvider>
    {children}
  </ManorProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: ManorTestWrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
