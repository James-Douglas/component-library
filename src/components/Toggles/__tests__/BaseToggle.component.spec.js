import React from 'react';
import { render, fireEvent } from '../../../testUtils';
import 'jest-styled-components';
import BaseToggle, { getInlineStyles } from '../BaseToggle';

describe('getInlineStyles()', () => {
  it('returns empty object when type is not rectangle', () => {
    expect(getInlineStyles('square', {})).toEqual({});
  });
  it('returns inline styles when type is rectangle without col', () => {
    const rectOptions = {
      height: 4,
    };

    const result = getInlineStyles('rectangle', rectOptions);
    expect(result.height).toEqual('4rem');
  });
  it('returns inline styles when type is rectangle with col', () => {
    const rectOptions = {
      height: 4,
      col: 2,
    };

    const result = getInlineStyles('rectangle', rectOptions);
    expect(result.height).toEqual('4rem');
    expect(result.width).toEqual('50%');
    expect(result.flexBasis).toEqual('50%');
  });
});

describe('BaseToggle', () => {
  it('renders with minimal props', () => {
    const { container } = render(<BaseToggle type="square" id="test-square" value="test" handleToggle={() => {}} />);
    expect(container).toMatchSnapshot();
  });
  it('sets checked when selectedId equals id', () => {
    const { container } = render(<BaseToggle type="square" id="test-selected-id" value="test" selectedValue="test" autofill handleToggle={() => {}} />);
    expect(container.querySelector('input')).toHaveAttribute('checked');
  });
  it('calls onToggle when toggled', () => {
    const onToggleCb = jest.fn();
    const { container } = render(<BaseToggle type="square" id="test-square" value="testt" handleToggle={onToggleCb} />);
    const element = container.firstChild;
    fireEvent.click(element);
    expect(onToggleCb).toHaveBeenCalled();
    expect(onToggleCb.mock.calls[0][0]).toEqual('testt');
  });
  it('calls focus and blur handlers', () => {
    const handleFocus = jest.fn();
    const handleBlur = jest.fn();
    const { container } = render(<BaseToggle type="square" handleToggle={() => {}} id="test" value="testt" handleFocus={handleFocus} handleBlur={handleBlur} />);
    const input = container.querySelector('input');
    fireEvent.focus(input);
    expect(handleFocus).toHaveBeenCalled();
    fireEvent.blur(input);
    expect(handleBlur).toHaveBeenCalled();
  });
});
