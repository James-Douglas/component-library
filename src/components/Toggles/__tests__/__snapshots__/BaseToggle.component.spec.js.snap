// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseToggle renders with minimal props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  cursor: pointer;
  border: 1px solid #DDDDDD;
  font-weight: 700;
  margin: 0 0.8rem 1.6rem 0.8rem;
  background: #FFFFFF;
  box-shadow: 0 0.5rem 0.5rem 0 rgba(0,0,0,.1);
  line-height: 1.5;
  min-width: 12.8rem;
}

.c1 {
  opacity: 0;
  width: 0;
  height: 0;
  position: absolute;
  fill: currentColor;
  box-shadow: none;
  -webkit-transition: all 200ms ease-out;
  transition: all 200ms ease-out;
}

.c1:checked + label svg {
  color: #FFFFFF;
}

.c1:disabled + label:hover {
  cursor: not-allowed;
}

.c1:disabled + label {
  color: #CCCCCC;
  border: 1px solid transparent;
  fill: currentColor;
}

.c1:checked + label {
  background: #001443;
  color: #FFFFFF;
  box-shadow: none;
}

.c1:focus + label {
  box-shadow: 0 0 2px 3px rgba(0,123,255,.3);
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      id="test-square"
      name=""
      tabindex="0"
      type="radio"
      value="test"
    />
  </div>
</div>
`;
